File: interview/algorithm/test_algorithms/test_algorithms/longest_increasing_sequence.cpp
#include <stdio.h>
#include <assert.h>
#include <string.h>

static int DP(const int* seq, const int nseq, int i, int* num_cached, int* prev_nodes)
{
	assert(i >= 0 && i < nseq);

	if (num_cached[i] == -1)
	{
		int prev_node = -1;
		int num_res = 1;

		if (i == 0)
		{
			num_res = 1;
		}
		else
		{
			for (int jj = 0; jj < i; jj++)
			{
				if (seq[i] > seq[jj])
				{
					int dp = DP(seq, nseq, jj, num_cached, prev_nodes) + 1;
					if (dp > num_res)
					{
						prev_node = jj;
						num_res = dp;
					}
				}
			}
		}

		num_cached[i] = num_res;
		prev_nodes[i] = prev_node;
	}

	assert(num_cached[i] != -1);
	return num_cached[i];
}

static int solution(const int* seq, const int nseq)
{
	int num_cached[64];
	int prev_nodes[64];
	memset(num_cached, -1, sizeof(num_cached));
	memset(prev_nodes, -1, sizeof(prev_nodes));
	int num = DP(seq, nseq, nseq - 1, num_cached, prev_nodes);
	return num;
}

void test_longest_increasing_sequence()
{
	//int seq[] = { 2, 4, 3, 5, 1, 7, 6, 9, 8 };
	int seq[] = { 4, 5, 1, 2, 3, };
	//int seq[] = { 4, 1, 2, };
	//int seq[] = { 2, 4, 3, 5 };
	int nlen = sizeof(seq) / sizeof(seq[0]);
	int val = solution(seq, nlen);
	printf("val: %d\n", val);
}

File: interview/test_code/longest_increasing_sequence.py

def DP(seq, i):
    nseq = len(seq)

    if i == 0:
        res = list()
        res.append(seq[i])
        return res

    best_num = 1
    ret_list = list()
    ret_list.append(seq[i])

    for j in range(0, i):
        if seq[i] > seq[j]:
            j_res = DP(seq, j)
            num = len(j_res) + 1

            if num > best_num:
                best_num = num
                ret_list = j_res.copy()
                ret_list.append(seq[i])

    return ret_list


def solution(seq):
    nseq = len(seq)
    return DP(seq, nseq - 1)

def test_longest_increasing_sequence():
    seq = [ 2, 4, 3, 5, 1, 7, 6, 9, 8 ]
    # seq = [ 4, 5, 1, 2, 3 ]
    # int seq[] = { 4, 1, 2, };
    # int seq[] = { 2, 4, 3, 5 };
    val = solution(seq)
    print(val)

test_longest_increasing_sequence()